tools/distpartitioning/dist_lookup.py

```diff
@@ -0,0 +1,289 @@
+            #slice the partition-ids which belong to the current instance
+            partid_list.append(ntype_partids[start:end])
+
+        #store all the information in the object instance variable
+        self.id_map = id_map
+        self.type_nid_begin = np.array(type_nid_begin, dtype=np.int64)
+        self.type_nid_end = np.array(type_nid_end, dtype=np.int64)
+        self.partid_list = partid_list
+        self.ntype_count = np.array(ntype_count, dtype=np.int64)
+        self.rank = rank
+        self.world_size = world_size
+
+    def get_partition_ids(self, global_nids):
+        '''
+        This function is used to get the partition-ids for a given set of global node ids
+
+        global_nids <-> partition-ids mappings are deterministically  distributed across 
+        all the participating processes, within the service. A contiguous global-nids
+        (ntype-ids, per-type-nids) are stored within each process and this is determined 
+        by the total no. of nodes of a given ntype-id and the rank of the process.
+
+        Process, where the global_nid <-> partition-id mapping is stored can be easily computed
+        as described above. Once this is determined we perform an alltoallv to send the request.
+        On the receiving side, each process receives a set of global_nids and retrieves corresponding
+        partition-ids using locally stored lookup tables. It builds responses to all the other
+        processes and performs alltoallv.
+
+        Once the response, partition-ids, is received, they are re-ordered corresponding to the 
+        incoming global-nids order and returns to the caller.
+
+        Parameters:
+        -----------
+        self : instance of this class
+            instance of this class, which is passed by the runtime implicitly
+        global_nids : numpy array
+            an array of global node-ids for which partition-ids are to be retrieved by 
+            the distributed lookup service.
+
+        Returns:
+        --------
+        list of integers : 
+            list of integers, which are the partition-ids of the global-node-ids (which is the
+            function argument)
+        '''
+
+        #find the process where global_nid --> partition-id(owner) is stored. 
+        ntype_ids, type_nids = self.id_map(global_nids)
+        ntype_ids, type_nids = ntype_ids.numpy(), type_nids.numpy()
+        assert len(ntype_ids) == len(global_nids)
+
+        #For each node-type, the per-type-node-id <-> partition-id mappings are
+        #stored as contiguous chunks by this lookup service. 
+        #The no. of these mappings stored by each process, in the lookup service, are
+        #equally split among all the processes in the lookup service, deterministically
+        service_owners = self.ntype_count[ntype_ids]
+        service_owners = np.ceil(service_owners/self.world_size).astype(np.int64)
+        service_owners = np.floor_divide(type_nids, service_owners).astype(np.int64)
```

Comment:
maybe it's better to write them together. the way it writes may not be very intuitive.

Reply:
Changed the variable names to appropriately represent the values stored in these variable names. This should remove some of the confusion.